test/controllers/tasks_controller_test.rb

```diff
@@ -79,50 +74,105 @@
 
       must_respond_with :redirect
       must_redirect_to task_path(new_task.id)
     end
   end
 
-  # Unskip and complete these tests for Wave 3
+
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+      # Act
+      get edit_task_path(task.id)
+
+      # Assert 
+      must_respond_with :success
     end
 
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+
+      get edit_task_path(-1)
+
+      must_respond_with :redirect
+      expect(flash[:error]).must_equal "That task does not exist."
     end
   end
 
-  # Uncomment and complete these tests for Wave 3
+
   describe "update" do
-    # Note:  If there was a way to fail to save the changes to a task, that would be a great
-    #        thing to test.
     it "can update an existing task" do
-      skip
-      # Your code here
+      new_task = Task.create!(name: 'Household', description: 'Mop the floor', completion_date: "April 15th")
+      
+      new_name = { 
+        task: { 
+          name: 'Changed name', 
+          } 
+        }
+
+      patch task_path(new_task.id), params: new_name
+      new_task.reload
+      expect(new_task.name).must_equal "Changed name"
     end
 
+
     it "will redirect to the root page if given an invalid id" do
-      skip
-      # Your code here
+      patch task_path(-1)
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
     end
   end
 
-  # Complete these tests for Wave 4
   describe "destroy" do
-    # Your tests go here
+    it 'deletes a task completely from the database' do 
+      task = Task.create!(name: "Home", description: "Do the dishes")
+
+      expect {
+        delete task_path(task)
+      }.must_change "Task.count", -1
+
+      must_respond_with :redirect
+      must_redirect_to tasks_path
 
+      deleted_task = Task.find_by(id: task.id)
+      expect(deleted_task).must_be_nil
+    end
   end
 
-  # Complete for Wave 4
+  # This test could be far more thorough
   describe "toggle_complete" do
-    # Your tests go here
+    it "returns success" do
+      patch complete_task_path(task.id)
+
+      must_respond_with :found
+    end
+
+    it "updates completion date to current date" do
+      patch complete_task_path(task.id)
+
+      complete = Task.find_by(id: task.id)
+
+      expect(complete.completion_date).must_equal (DateTime.now).to_s
+    end
+
+    # Hi Chris! I tried to get this test to Pass but I couldn't manage!! Help please! :)
+
+    # it "won't update completed task date during complete incomplete toggle" do
+    #   task = Task.create(
+    #     name: "Home",
+    #     description: "Walk the dog",
+    #     completion_date: "April, 11th 2019",
+    #   )
+
+    #   patch complete_task_path(task.id)
+
+    #   toggle = Task.find_by(id: task.id)
+    
+    #   expect(toggle.completion_date).must_equal "April, 11th 2019"
```

Comment:
This is going to need to do something like:
```ruby
      expect(DateTime.parse(toggle.completion_date).to_i).must_be_close_to  DateTime.now.to_i, 5
```

This is because your TasksController is saving a DateTime into the string field.  

Note you should also test the `completed` field